source/slang/slang.cpp

```diff
@@ -623,6 +623,17 @@ SLANG_NO_THROW slang::TypeLayoutReflection* SLANG_MCALL Linkage::getTypeLayout(
     return asExternal(typeLayout);
 }
 
+SLANG_NO_THROW SlangResult SLANG_MCALL Linkage::getTypeConformanceWitnessMangledName(
+    slang::TypeReflection* type, slang::TypeReflection* interfaceType, ISlangBlob** outNameBlob)
+{
+    auto subType = asInternal(type);
+    auto supType = asInternal(interfaceType);
+    auto name = getMangledNameForConformanceWitness(subType->getASTBuilder(), subType, supType);
+    Slang::ComPtr<ISlangBlob> blob = Slang::StringUtil::createStringBlob(name);
+    *outNameBlob = blob.detach();
+    return SLANG_OK;
```

Comment:
Wait a second... This logic isn't actually validating that the given type actually does conform to the interface. It is just trying to return what the mangled name for the conformance *would be* if it exists, and leaves it as a linker error if/when the conformance doesn't actually exist.

I'm worried that this isn't the right way to go about things.

Conceptually, the Right Way to handle this would be to call into the AST/semantics level with a `doesTypeConformToInterface` or `getSubtypeWitness` or whatever the operation is called, to either produce a `SubtypeWitness*` value, or `nullptr`. If a subtype witness can be found, then we can get its mangled name directly and we are happy/done (although at that point we can also just return the `SubtypeWitness*` as a reified object that represents the conformance.

Such a lookup needs to be done with specific modules (and their `ExtensionDecl`s) in scope, so it would need to be an operation on the `ISession` (`Linkage`).

If we want to have an API that eventually interacts with linking, then the user needs a way to say "my compiled program needs to include the code for the conformance of `MyType` to `IFoo`." It may be that the conformance is *not* in the module that defines `MyType`, nor the module that defines `IFoo`, but is instead in some third module `M`.

Just because we can form the mangled name that represents a `MyType : IFoo` conformance, that does not do anything to help us identify the module `M`, and when it comes time to link together the code for our program, nothing about `MyType`, `IFoo`, or that mangled name will help ensure we link in the code from `M`.

What the user needs is a way to get a handle to the `MyType : IFoo` conformance (which is a `SubtypeWitness`) and to specify that the conformance represented by that witness had better be linked in. Inside the compiler, that witness eventually refers to an `InheritanceDecl`, and we simply need to ensure that the given `InheritanceDecl` (and the module it comes from) show up in the linked code.

I'm coming away convinced that this API *needs* to be written in terms of an interaction with the compiler front-end to do proper semantic checking of the conformance (which will yield a `SubtypeWitness`).

I'm fine with this as a short-term hack, but I don't think it should be the basis for our longer-term API for this feature.